### PR TITLE
New function createAsset

### LINKED ISSUE #483: Import Base64-encoded image data to Asset
**Is your feature request related to a problem? Please describe.**
I am working on a project to import characters from Chummer5a. Getting the data works fine so far but I would like to have the option to get the images provided by Chummer5a as URL handeled by MT. Or as Phergus put it: "What is missing on our end is the code to grab the image given that URL (or the Base64 data) and then cache it locally like the normal image assets MT uses.  Not a huge deal, just someone needs to get a chance to do the work."

**Describe the solution you'd like**
Code to grab the image and then cache it locally like the normal image assets MT uses.

**Additional context**
See discord -> #support4dev 15.05.19

### PR DESCRIPTION
**Pull request by Jmr3366**
### Description of the Change
- Resolves #483 
https://discord.com/channels/296230822262865920/296658523725496320/578194944871759909

New trusted macro function `createAsset(name,imageData)` that will return an `asset:\\< asset ID >` value on success.
Checks are made for data types and only `webp`,`jpg`,`png` are permitted data types.

### Possible Drawbacks
- `createAsset` can allow scriptable asset creation and if implemented improperly, _could be_ campaign breaking.
With great power comes great responsibility.

### Release Notes
Examples: 
`[r: createAsset("Test", < base64 encoded image data >)]`
`[r: createAsset("Test", "https://website.com/image.webp")]`
outputs 
`asset:\\< asset ID >`

(I renamed my branch as I almost overwrote it)

- - -
This change is [Reviewable](https://reviewable.io/reviews/RPTools/maptool/4923)

### CODE REVIEW

**Comment by cwisniew** [comment]
This needs to ensure the asset is valid before adding it to the campaign

**Comment by Jmr3366** [comment]
I do check file headers for the 3 file formats this function supports.  If there are other checks I can do or add, I will try to comply with those.

File: src/main/java/net/rptools/maptool/client/functions/TokenImage.java [hunk 1]
```diff
@@ -171,6 +173,34 @@ public Object childEvaluate(
...
+      String imageString = args.get(1).toString();
+      if (imageName == "" || imageString == "") {
+        throw new ParserException(
+            I18N.getText("macro.function.general.paramCannotBeEmpty", functionName));
+      } else {
+        if (imageString.length() > 8) {
+          byte[] imageBytes = Base64.decode(imageString);
+          String imageCheck = new String(imageBytes, 0, 4);
+          /* header check for: webp || jpg || png */
+          if (imageCheck.equals("RIFF") || imageCheck.equals("ÿØÿà") || imageCheck.equals("‰PNG")) {
```
Comment (cwisniew) [inline src/main/java/net/rptools/maptool/client/functions/TokenImage.java]: If you make these constants so its more meaningful when people read the code I will approve it :)

**Review by cwisniew — CHANGES_REQUESTED** [review]

File: src/main/java/net/rptools/maptool/client/functions/TokenImage.java (hunk 1 above)
Comment (Jmr3366) [reply src/main/java/net/rptools/maptool/client/functions/TokenImage.java]: Is this acceptable or _inline_ with your thoughts?
```
public static final String FILE_HEADER_WEBP = "RIFF";
public static final String FILE_HEADER_JPG = "ÿØÿà";
public static final String FILE_HEADER_PNG = "‰PNG";
  ....
if (imageCheck.equals(FILE_HEADER_WEBP) || imageCheck.equals(FILE_HEADER_JPG) || imageCheck.equals(FILE_HEADER_PNG)) {```

edit - double posted . .  .

Comment (Jmr3366) [reply src/main/java/net/rptools/maptool/client/functions/TokenImage.java:189]: Is this in-line with what you're wanting or am I _off_ :) ?
``` if (imageCheck.equals(FILE_HEADER_WEBP) || imageCheck.equals(FILE_HEADER_JPG) || imageCheck.equals(FILE_HEADER_PNG)) {```

Comment (Jmr3366) [reply src/main/java/net/rptools/maptool/client/functions/TokenImage.java]: spotless, I hate thee . . .
Added the constants, better checks re: empty string and npe catch
Added "https:" support for the same file types.

**Review by cwisniew — APPROVED** [review]

**Pull request merged**